app/controllers/application_controller.rb

```diff
@@ -202,12 +203,29 @@ def refresh_launchpad_if_needed
     # Launchpad login
     if launchpad_login_required?
       if logged_in? && server_session_expires_in < 0
-        # TODO until the keep alive is fully implemented (MMT-1297) we should just ask the user to login with launchpad again
-        redirect_to sso_url
+        result = refresh_launchpad
+
+        redirect_to sso_url if result[:error]
       end
     end
   end
 
+  def refresh_launchpad
+    return { error: 'session older than 8 hours' } if Time.now.to_i - session[:original_logged_in_at] > 28_800
+
+    response = cmr_client.keep_alive(token)
+    Rails.logger.info "launchpad integration keep alive endpoint response: #{response.inspect}" if Rails.env.development?
+    if response.success?
+      session[:launchpad_cookie] = response.headers.fetch('set-cookie', '').split("#{launchpad_cookie_name}=").last
```

Comment:
I had added `pull_launchpad_cookie` in `SamlController` in anticipation of needing to use it for this. We can leave this as is, but you could also use that as well.